plugins/base/src/main/kotlin/renderers/html/HtmlTemplater.kt

```diff
@@ -0,0 +1,90 @@
+package org.jetbrains.dokka.base.renderers.html
+
+import freemarker.cache.ClassTemplateLoader
+import freemarker.cache.FileTemplateLoader
+import freemarker.cache.MultiTemplateLoader
+import freemarker.log.Logger
+import freemarker.template.Configuration
+import freemarker.template.TemplateExceptionHandler
+import kotlinx.coroutines.async
+import kotlinx.coroutines.runBlocking
+import org.jetbrains.dokka.base.DokkaBase
+import org.jetbrains.dokka.base.DokkaBaseConfiguration
+import org.jetbrains.dokka.plugability.DokkaContext
+import org.jetbrains.dokka.plugability.configuration
+import java.io.StringWriter
+import java.util.concurrent.ConcurrentHashMap
+
+
+enum class DokkaTemplateTypes(val path: String) {
+    BASE("base.ftl")
+}
+
+typealias TemplateMap = Map<String, Any?>
+
+class HtmlTemplater(
+    context: DokkaContext
+) {
+
+    init {
+        // to disable logging, but it isn't reliable see [Logger.SYSTEM_PROPERTY_NAME_LOGGER_LIBRARY]
+        // (use SLF4j further)
+        System.setProperty(
+            Logger.SYSTEM_PROPERTY_NAME_LOGGER_LIBRARY,
+            System.getProperty(Logger.SYSTEM_PROPERTY_NAME_LOGGER_LIBRARY) ?: Logger.LIBRARY_NAME_NONE
+        )
+    }
+
+    private val configuration = configuration<DokkaBase, DokkaBaseConfiguration>(context)
+    private val templaterConfiguration =
+        Configuration(Configuration.VERSION_2_3_31).apply { configureTemplateEngine() }
+    private val cachedTemplates: MutableSet<DokkaTemplateTypes> =
+        ConcurrentHashMap<DokkaTemplateTypes, Boolean>().keySet(true)
+
+
+    private fun Configuration.configureTemplateEngine() {
+        val loaderFromResources = ClassTemplateLoader(javaClass, "/dokka/templates")
+        templateLoader = configuration?.templatesDir?.let {
+            MultiTemplateLoader(
+                arrayOf(
+                    FileTemplateLoader(it),
+                    loaderFromResources
+                )
+            )
+        } ?: loaderFromResources
+
+        unsetLocale()
+        defaultEncoding = "UTF-8"
+        templateExceptionHandler = TemplateExceptionHandler.RETHROW_HANDLER
+        logTemplateExceptions = false
+        wrapUncheckedExceptions = true
+        fallbackOnNullLoopVariable = false
+    }
+
+    fun setupSharedModel(model: TemplateMap) {
+        templaterConfiguration.setSharedVariables(model)
+    }
+
+    fun renderFromTemplate(
+        templateType: DokkaTemplateTypes,
+        generateModel: () -> TemplateMap
+    ): String {
+        val out = StringWriter()
+        // Freemarker has own cache to keep templates
+        if (cachedTemplates.contains(templateType)) { // it's a heuristic, freemarker can remove a template from cache
```

Comment:
> It seems like this will never be true? We only populate cachedTemplates if we're inside this if, but we'll never get in it because the cache is empty in the beginning. Am I missing something?

My bad.
`if (!cachedTemplates.contains(templateType))` should be.

> Also, just to point out, this is still not entirely "thread safe". Two threads can see cachedTemplates.contains as false and start populating it, one overriding the result of the other. Not a big deal, but I still don't understand why we need it :)

Yes, so it's  a heuristic.